[typedef] variable-declaration-ignore-function from TSLint not implemented

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/typedef": [ "error", {
            "arrayDestructuring": true,
            "arrowParameter": true,
            "memberVariableDeclaration": true,
            "objectDestructuring": true,
            "parameter": true,
            "propertyDeclaration": true,
            "variableDeclaration": true
      }]
  }
}
```

```TS
const doSomething = (): void => {
	console.log('Doing something');
};

doSomething();
```

**Expected Result**
No error from ESLint

**Actual Result**
```
...\src\index.ts
  1:7  error  expected doSomething to have a type annotation  @typescript-eslint/typedef
```
**Additional Info**
When I change the code to
```TS
const doSomething: () => void = (): void => {
	console.log('Doing something');
};

doSomething();
```
the code passes linting.
When I check the rule page of TSLint (https://palantir.github.io/tslint/rules/typedef/) I see they have an option to ignore variable declarations for a function.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.18.0` |
| `@typescript-eslint/parser`        | `2.18.0` |
| `TypeScript`                       | `3.7.5` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `13.7.0` |
| `npm`                              | `6.13.6` |

--- a/packages/eslint-plugin/docs/rules/typedef.md
+++ b/packages/eslint-plugin/docs/rules/typedef.md
@@ -37,7 +37,8 @@ This rule has an object option that may receive any of the following as booleans
 - `"objectDestructuring"`
 - `"parameter"`: `true` by default
 - `"propertyDeclaration"`: `true` by default
-- `"variableDeclaration"`
+- `"variableDeclaration"`,
+- `"variableDeclarationIgnoreFunction"`
 
 For example, with the following configuration:
 
@@ -246,6 +247,30 @@ Examples of **correct** code with `{ "variableDeclaration": true }`:
 const text: string = 'text';
 let initialText: string = 'text';
 let delayedText: string;
+```
+
+### `variableDeclarationIgnoreFunction`
+
+Ignore variable declarations for non-arrow and arrow functions.
+
+Examples of **incorrect** code with `{ "variableDeclaration": true, "variableDeclarationIgnoreFunction": true }`:
+
+```ts
+const text = 'text';
+```
+
+Examples of **correct** code with `{ "variableDeclaration": true, "variableDeclarationIgnoreFunction": true }`:
+
+```ts
+const a = (): void => {};
+const b = function (): void => {};
+const c: () => void = (): void => {};
+
+class Foo {
+  a = (): void => {};
+  b = function (): void => {};
+  c = () => void = (): void => {};
+}
 ```
 
 ## When Not To Use It

--- a/packages/eslint-plugin/src/rules/typedef.ts
+++ b/packages/eslint-plugin/src/rules/typedef.ts
@@ -12,6 +12,7 @@ const enum OptionKeys {
   Parameter = 'parameter',
   PropertyDeclaration = 'propertyDeclaration',
   VariableDeclaration = 'variableDeclaration',
+  VariableDeclarationIgnoreFunction = 'variableDeclarationIgnoreFunction',
 }
 
 type Options = { [k in OptionKeys]?: boolean };
@@ -41,6 +42,7 @@ export default util.createRule<[Options], MessageIds>({
           [OptionKeys.Parameter]: { type: 'boolean' },
           [OptionKeys.PropertyDeclaration]: { type: 'boolean' },
           [OptionKeys.VariableDeclaration]: { type: 'boolean' },
+          [OptionKeys.VariableDeclarationIgnoreFunction]: { type: 'boolean' },
         },
       },
     ],
@@ -100,6 +102,14 @@ export default util.createRule<[Options], MessageIds>({
       }
     }
 
+    function isVariableDeclarationIgnoreFunction(node: TSESTree.Node): boolean {
+      return (
+        !!options[OptionKeys.VariableDeclarationIgnoreFunction] &&
+        (node.type === AST_NODE_TYPES.FunctionExpression ||
+          node.type === AST_NODE_TYPES.ArrowFunctionExpression)
+      );
+    }
+
     return {
       ArrayPattern(node): void {
         if (options[OptionKeys.ArrayDestructuring] && !node.typeAnnotation) {
@@ -112,6 +122,10 @@ export default util.createRule<[Options], MessageIds>({
         }
       },
       ClassProperty(node): void {
+        if (node.value && isVariableDeclarationIgnoreFunction(node.value)) {
+          return;
+        }
+
         if (
           options[OptionKeys.MemberVariableDeclaration] &&
           !node.typeAnnotation
@@ -155,7 +169,8 @@ export default util.createRule<[Options], MessageIds>({
           (node.id.type === AST_NODE_TYPES.ArrayPattern &&
             !options[OptionKeys.ArrayDestructuring]) ||
           (node.id.type === AST_NODE_TYPES.ObjectPattern &&
-            !options[OptionKeys.ObjectDestructuring])
+            !options[OptionKeys.ObjectDestructuring]) ||
+          (node.init && isVariableDeclarationIgnoreFunction(node.init))
         ) {
           return;
         }

--- a/packages/eslint-plugin/tests/rules/typedef.test.ts
+++ b/packages/eslint-plugin/tests/rules/typedef.test.ts
@@ -254,6 +254,57 @@ ruleTester.run('typedef', rule, {
         },
       ],
     },
+    // variable declaration ignore function
+    {
+      code: `const foo = function(): void {};`,
+      options: [
+        {
+          variableDeclaration: true,
+          variableDeclarationIgnoreFunction: true,
+        },
+      ],
+    },
+    {
+      code: `const foo = (): void => {};`,
+      options: [
+        {
+          variableDeclaration: true,
+          variableDeclarationIgnoreFunction: true,
+        },
+      ],
+    },
+    {
+      code: `const foo: () => void = (): void => {};`,
+      options: [
+        {
+          variableDeclaration: true,
+          variableDeclarationIgnoreFunction: true,
+        },
+      ],
+    },
+    {
+      code: `const foo: () => void = function (): void {};`,
+      options: [
+        {
+          variableDeclaration: true,
+          variableDeclarationIgnoreFunction: true,
+        },
+      ],
+    },
+    {
+      code: `
+class Foo {
+  a = (): void => {};
+  b = function (): void {};
+}
+      `,
+      options: [
+        {
+          variableDeclaration: true,
+          variableDeclarationIgnoreFunction: true,
+        },
+      ],
+    },
   ],
   invalid: [
     // Array destructuring
@@ -601,6 +652,75 @@ ruleTester.run('typedef', rule, {
       options: [
         {
           variableDeclaration: true,
+        },
+      ],
+    },
+    {
+      code: `const foo = 'foo'`,
+      errors: [
+        {
+          messageId: 'expectedTypedefNamed',
+          data: { name: 'foo' },
+        },
+      ],
+      options: [
+        {
+          variableDeclaration: true,
+          variableDeclarationIgnoreFunction: true,
+        },
+      ],
+    },
+    {
+      code: `const foo = function(): void {};`,
+      errors: [
+        {
+          messageId: 'expectedTypedefNamed',
+          data: { name: 'foo' },
+        },
+      ],
+      options: [
+        {
+          variableDeclaration: true,
+          variableDeclarationIgnoreFunction: false,
+        },
+      ],
+    },
+    {
+      code: `const foo = (): void => {};`,
+      errors: [
+        {
+          messageId: 'expectedTypedefNamed',
+          data: { name: 'foo' },
+        },
+      ],
+      options: [
+        {
+          variableDeclaration: true,
+          variableDeclarationIgnoreFunction: false,
+        },
+      ],
+    },
+    {
+      code: `
+class Foo {
+  a = (): void => {};
+  b = function (): void {};
+}
+      `,
+      errors: [
+        {
+          messageId: 'expectedTypedefNamed',
+          data: { name: 'a' },
+        },
+        {
+          messageId: 'expectedTypedefNamed',
+          data: { name: 'b' },
+        },
+      ],
+      options: [
+        {
+          variableDeclaration: true,
+          variableDeclarationIgnoreFunction: false,
         },
       ],
     },

